1.16.x Documentation Start

ChampionAsh5357 opened the pull request:
This contains about a 75%-80% update to the current forge docs written for 1.16.x as of update 32.0.38. I accidentally pulled the 1.14.x instead of the 1.15.x for a base so there might be some commits from others or information that I forgot to remove. The main issues such as updating methods and fields to their current counterparts is approximately done say for a few pages which I will discuss later.

I know I am merging this to the 1.15.x docs. There isn't one for 1.16.x as of yet.

**Format Standards**
There was a lot of formatting standards that need to be addressed and probably still do, specifically dealing with the use of `Class/Interface Names`. Some methods are referenced via the class/interface they extended/implemented instead of the class they were implemented (e.g. `IForgeBakedModel` and `IBakedModel`). I would like to know in which cases the method should be referenced from which class/interface.

**Removal**
I have removed a lot of text pertaining to 1.8 and 1.9 mentions as the system prior is almost completely inapplicable in the current standard. There are still a few mentions left which I think should be redone to mention the current system in slightly more depth. All other text removals have to do with a change in implementation.

**Obfuscation**
There are a few obfuscated methods scattered around the documentation due to its recent release. I will update them as new mappings get released.

**BlockState**
Currently, the documentation for BlockState uses DyeColor for a PropertyEnum. As of 1.13, the flattening removed the use of this property since the block registry was now expressively infinite. For this reason, I replaced the use of DyeColor as an example. In regards to the subset example on PropertyEnum, DyeColor still remains the best example for the subset over all current enums. So, I might have to create a specific example to show this change without creating a property that should be expressed as different blocks.

Also, the documentation specifies overwriting getDefaultState(). I do believe this should be specified to calling setDefaultState(baseState) within the constructor to better clarify. I did not change this yet as I wanted the opinion of the reviewer.

**Model**
I could not accurately express the model page when converting from the current language had many mappings  within the current system. I did my best job to try and convert the names successfully. A deep review is needed to make sure there is nothing I missed or need to fix.

**Animation and WorldSavedData**
These two systems have not been well documented since their 1.12 versions. I have done very little editing on them since I still need to verify how to work the current implementation within 1.16. Until then, I will leave them untouched until I am more fluent with them.

Please let me know if there is anything I can do to improve the current condition of the docs. As this version is incomplete, it still provides a better understanding of the current system in place over what the docs has right now.

TheGreyGhost committed:
Updated Environment Setup instructions for IntelliJ (#260)

* Updated Environment Setup instructions for IntelliJ

The gradle tasks have been changed since the original instructions were written.  IDEA 2019 also performs most of the setup without manual tweaks required.

* Updated in response to pull request review

ChampionAsh5357 committed:
Update structuring.md

ChampionAsh5357 committed:
Updated to 1.16.x

Block$Properties replaced with AbstractBlock$Properties

lightValue is currently unmapped and now takes a function with a BlockState parameter that returns an integer between 0-15.

ChampionAsh5357 committed:
Blocks Update

Fixed names and standardized some format.

EnumProperty explanation uses DyeColor enum as an example for subsets. The Flattening makes it pointless to use this as a property since all blocks should just be created instead. No better example enum however.

Should specify setDefaultState(this.baseState) over overwriting getDefaultState()

ChampionAsh5357 committed:
Tile Entities Update

Fix names and current mappings.

Need check on correct descriptions for TER#render

ChampionAsh5357 committed:
Format Fix

ChampionAsh5357 committed:
More fixes.

Should rewrite BlockState introduction entirely to match current conditions.

Need to finish Animation after testing implementations.

ChampionAsh5357 committed:
Model Implementation Update

Should fix some choice wording having to do with the old model system.

Should menetion that handlePerspective is wrapped via IForgeBakedModel and not IBakedModel due to language in other docs.

Names that are obfuscated are not mapped as of 32.0.38.

ChampionAsh5357 committed:
Finishing Port to 1.16.x

Textures are now handled in block/item instead of blocks/items

Capability converted to current implementation, could use a second read.

WorldSavedData still needs a major overhaul.

mrp-v2 commented on docs/datastorage/capabilities.md:
```diff
@@ -43,27 +43,15 @@ There's two ways to obtain such an instance, through the Capability itself, or b
-    return true;
+public <T> LazyOptional<T> getCapability(Capability<T> cap, Direction side) {
+  if (cap == CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) {
+    return (T) inventoryHandler.cast();
```
```suggestion
    return LazyOptional.of(inventoryHandler).cast();
```
inventoryHandler needs to be passed to LazyOptional.of() and then cast.

LexManos replied:
DO NOT Return a new LazyOptional every time -.-

mrp-v2 commented on docs/datastorage/capabilities.md:
````diff
@@ -43,27 +43,15 @@ There's two ways to obtain such an instance, through the Capability itself, or b
+Once you have your own instance of the capability interface, you will want to notify users of the capability system that you expose this capability and provide an optional of the interface reference. This is done by overriding the `getCapability` method, and comparing the instance with the capability you are exposing. If your machine has different slots based on which side is being queried, you c ...
 
 ```Java
 @Override
````
```suggestion
//somewhere in your TileEntity subclass
LazyOptional<IItemHandler> inventoryHandlerLazyOptional;
 
 //after initializing inventoryHandler
inventoryHandlerLazyOptional = LazyOptional.of(() -> inventoryHandler);

@Override
```

ChampionAsh5357 replied:
Added the clarity of LazyOptional in the recent commit

mrp-v2 commented on docs/datastorage/capabilities.md:
```diff
@@ -43,27 +43,15 @@ There's two ways to obtain such an instance, through the Capability itself, or b
-    return true;
+public <T> LazyOptional<T> getCapability(Capability<T> cap, Direction side) {
+  if (cap == CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) {
+    return (T) inventoryHandler.cast();
```
```suggestion
    return inventoryHandlerLazyOptional.cast();
```

ChampionAsh5357 replied:
Updated the TileEntity::getCapability method to support the LazyOptional example created above.

ChampionAsh5357 committed:
Capability LazyOptional Clarity Example

Updated capability attachment in TileEntity to better clarify it's tie in with LazyOptional as specified by mrp-v2

mrp-v2 committed:
Change onBlockActivated to new return type

Added some explanation for using `ActionResultType`.

mrp-v2 committed:
change example usage to a table

mrp-v2 committed:
provide example register methods

mrp-v2 committed:
fix formatting on example methods

ChampionAsh5357 committed:
Merge pull request #2 from mrp-v2/patch-2

Example methods for color model handlers.

ChampionAsh5357 committed:
Merge pull request #1 from mrp-v2/patch-1

Change onBlockActivated to new return type

Yunus1903 commented:
Might want to update the versions in [docs/legacy/index.md](https://github.com/MinecraftForge/Documentation/blob/1.15.x/docs/legacy/index.md)
And maybe delete (or update) [docs/legacy/porting1214.md](https://github.com/MinecraftForge/Documentation/blob/1.15.x/docs/legacy/porting1214.md)

ChampionAsh5357 committed:
Fixes suggested by Yunus1903

Accuracy of 1.14/1.15 docs need review

Readd porting once stable and MMMS for mappings.

ChampionAsh5357 commented:
> Might want to update the versions in [docs/legacy/index.md](https://github.com/MinecraftForge/Documentation/blob/1.15.x/docs/legacy/index.md)
> And maybe delete (or update) [docs/legacy/porting1214.md](https://github.com/MinecraftForge/Documentation/blob/1.15.x/docs/legacy/porting1214.md)

After a bit of review, I have updated the versions in the legacy index and removed the porting tab. The versions' accuracy is slightly better than 1.13.x, so I improved it to 15% each. However, these numbers can be adjusted after a more in-depth review of the current docs. As for the choice to delete porting from 1.12-1.14 instead of doing a 1.12-1.16, there is current no written primer, and the jump would go through at least two or three extra links for transferring over. Since 1.12 hasn't been support for a while and the current LTS is 1.15, this should be readded once a stable build has been released with the introduction of the MMMS. From there, a detailed primer can be written.

I have also updated the sidebar navigation to reflect the changes.

Yunus1903 commented on mkdocs.yml:
```diff
@@ -84,7 +82,7 @@ theme:
 
 extra:
-  current_version: 1.14.x
+  current_version: 1.16.x
```
Not sure what this defines but I asked tterrag earlier about the version that shows up when you open the docs and he said it should be the LTS version. So if this is that, it should be 1.15.x

ChampionAsh5357 replied:
If that's what it is, I'll update date it to LTS instead of current.

ChampionAsh5357 committed:
current_version updated to LTS 1.15

tterrag1098 commented:
This is a bit of a shotgun PR, a lot of these fixes are applicable to 1.15 and 1.14 too, such as the class name fixes. I suggest splitting this up into a couple more focused PRs.

ChampionAsh5357 commented:
> This is a bit of a shotgun PR, a lot of these fixes are applicable to 1.15 and 1.14 too, such as the class name fixes. I suggest splitting this up into a couple more focused PRs.

I mean yeah. This was to address this issues surrounding the lack of updates for 1.13+ docs. In my opinion, all documentation should be actively updated and improved upon at all times when existing information is present. Since 1.14 was going out and 1.15 was becoming the new LTS, I just figured I would create an updated baseline instead of copy/paste the old docs into the new ones with relatively little update.

If you would like me to, I can leave this PR open and update 1.14.x and 1.15.x to this standard baseline as well. I'll probably do another 'shotgun PR' for both since updating all the information to the current standard is my goal and not adding new information to the docs. Or, if you would prefer, I can close this PR and split it into multiple PRs for each page/version. I'm fine with doing either as long as I am able to provide accurate and updated information. I would just like to see the docs provide 100% accurate information across the version once again.

tterrag1098 commented:
I need to start from 1.14.x so that fixes can be merged forward. Changes made directly to the newer branches should only contain things related to those versions *specifically*. You don't need to repeat the same changes across multiple branches.

ChampionAsh5357 commented:
> I need to start from 1.14.x so that fixes can be merged forward. Changes made directly to the newer branches should only contain things related to those versions _specifically_. You don't need to repeat the same changes across multiple branches.

Thank you very much for the information. I will update the docs in 1.14.x and 1.15.x accordingly to their newest state to comply. I should have a PR for each by tomorrow at the latest. I will keep this PR open until then and reclassify the information in this afterwards to also comply with those changes. 

In the case of model animations and worldsaveddata, should I just currently remove references to them until they can be verified and updated accordingly?

ChampionAsh5357 commented:
Documentation has been split into two separate PRs: one for 1.14.x (#322) and one for 1.15.x (#323). The 1.15.x PR only has changes specific to the version and no changes included within 1.14.x as requested. I will be redoing this PR to account for this change. This PR will then be closed and reestablished with only 1.16.x specific changes once I am done.

ChampionAsh5357 commented:
This has now been superseded by #324 to meet with tterrag1098 standards.
